ImmuDB4Net/ImmuClientSync.cs

```diff
@@ -0,0 +1,1364 @@
+            heartbeatCloseRequested = new ManualResetEvent(false);
+            heartbeatCalled = new ManualResetEvent(false);
+            StartHeartbeat();
+        }
+        finally
+        {
+            Interlocked.Exchange(ref sessionSetupInProgress, 0);
+        }
+    }
+
+    public void Reconnect()
+    {
+        lock (connectionSync)
+        {
+            ConnectionPool.Release(connection);
+            connection = ConnectionPool.Acquire(new ConnectionParameters
+            {
+                Address = GrpcAddress,
+                ShutdownTimeout = ConnectionShutdownTimeout
+            });
+        }
+    }
+
+    public void Close()
+    {
+        try
+        {
+            using (ManualResetEvent mre = new ManualResetEvent(false))
+            {
+                while (true)
+                {
+                    if (Interlocked.Exchange(ref sessionSetupInProgress, 1) == 0)
+                    {
+                        break;
+                    }
+                    mre.WaitOne(2);
+                }
+            }
+            StopHeartbeat();
+            SessionManager.CloseSession(Connection, ActiveSession);
+            ActiveSession = null;
+            lock (connectionSync)
+            {
+                ConnectionPool.Release(connection);
+                connection = releasedConnection;
+            }
+        }
+        finally
+        {
+            Interlocked.Exchange(ref sessionSetupInProgress, 0);
+        }
+    }
+
+    public bool IsClosed()
+    {
+        return Connection.Released;
+    }
+
+    private void CheckSessionHasBeenOpened()
+    {
+        if (ActiveSession == null)
+        {
+            throw new ArgumentException("Session is null. Make sure you call Open beforehand.");
+        }
+    }
+
+    private object stateSync = new Object();
+
+    public ImmuState State()
+    {
+        lock (stateSync)
+        {
+            ImmuState? state = stateHolder.GetState(ActiveSession, currentDb);
+            if (state == null)
+            {
+                state = CurrentState();
+                stateHolder.SetState(ActiveSession!, state);
+            }
+            else
+            {
+                CheckSessionHasBeenOpened();
+            }
+            return state;
+        }
+    }
+
+    /**
+    * Get the current database state that exists on the server.
+    * It may throw a RuntimeException if server's state signature verification fails
+    * (if this feature is enabled on the client side, at least).
+*/
+    public ImmuState CurrentState()
+    {
+        CheckSessionHasBeenOpened();
+        ImmudbProxy.ImmutableState state = Service.CurrentState(new Empty(), Service.GetHeaders(ActiveSession));
+        ImmuState immuState = ImmuState.ValueOf(state);
+        if (!immuState.CheckSignature(serverSigningKey))
+        {
+            throw new VerificationException("State signature verification failed");
```

Comment:
Suggestion: This could be a separate exception derived from VerificationException - that way it could be easier to check if the public signing key is invalid (instead of state mismatch for example)